### ISSUE TITLE
Investigate potential bug in how Authorizers of DTOs are resolved

### ISSUE THREAD

**Issue by lukasender** [doug-martin/nestjs-query#1344]
**Describe the bug**

`authorize` of DTOs involved in a query are not called in all cases, potentially leaking data to unauthorized users.

### Have you read the [Contributing Guidelines](https://github.com/doug-martin/nestjs-query/blob/master/CONTRIBUTING.md)?

yes

**To Reproduce**
1. I have prepared a **private repo** with a **working example** (derived from the `examples/auth` of this repo). It would be awesome if some of the core devs or anyone else who's willing to take some time and investigate/discuss the problem with me. I'd then **invite** you to [my private project](https://github.com/lukasender/nestjs-query-authorization/) and show you [my prepared example/pull request](https://github.com/lukasender/nestjs-query-authorization/pull/1) and step-by-step introductions on how to reproduce the problem.
2. I'd like to discuss the issue/potential bug there first **before** disclosing the behavior publicly (to prevent any abuse of the behavior).
3. I've already reached out to @doug-martin via email but did not get a response yet (I assume the email has gone to the spam folder or Doug just didn't get to it yet - we're all busy and bombarded with notifications. So, no offense taken here).
4. I've prepared a potential fix, but don't know the code base well enough to know if that's the correct solution.

**Expected behavior**
`authorize` of DTOs involved in a query are called in all cases

**Screenshots**
Will be available in my private repository/pull request

**Desktop (please complete the following information):**
 - Node Version: v14.17.6
 - Nestjs-query Version 88d744a7eb7f6b516d4914fee4cedd4773de82a4

**Comment by steinroe** [comment]
While investigating and fixing https://github.com/doug-martin/nestjs-query/issues/1336 (PR upcoming in the next few days), I have a decent understanding of the authorisation. Would be happy to provide some feedback 👍

**Comment by lukasender** [comment]
@steinroe thanks a lot! 🙏

Invite sent. There's only one Pull Request in that repo. I hope it explains everything with enough detail.

If not, feel free to leave questions directly on that PR.

**Comment by steinroe** [comment]
I guess this can be closes @lukasender 👍

**Comment by lukasender** [comment]
Yes, thanks a lot again!

The first test in the example repo verified it should be fixed https://github.com/doug-martin/nestjs-query/pull/1353#issuecomment-926434661

I'll be waiting for the new release and then test it with my bigger codebase.

If anything unexpected shows up, I'll report back again.

Thanks a lot, everyone! :)

**Comment by lukasender** [comment]
@steinroe @doug-martin I've tested the new version v0.30.0 and in my larger codebase and the problem still occurs.

I'm trying to fully understand the issue and come up with an example to prove the behavior (or better the unexpected behavior).

But so far, it seems I must reopen this issue.

Here's (an attempt at least) of what I've achieved to understand/debug so far. Maybe my assumption is wrong.

```js
@ObjectType('ResourceA')
@Authorize({
  authorize: () => {
    console.log('ResourceA: authorize called')
    return {}
  }
})
@Relation('ResourceB', () => ResourceBDto)
class ResourceADto {
  id: ...
}

// ----

@ObjectType('ResourceB')
@Authorize({
  authorize: () => {
    console.log('ResourceB: authorize called')
    return {}
  }
})
@Relation('ResourceC', () => ResourceCDto)
class ResourceBDto {
  id: ...
}

// ----

@ObjectType('ResourceC')
@Authorize({
  authorize: () => {
    console.log('ResourceC: authorize called')
    return {}
  }
})
class ResourceCDto {
  id: ...
}
```

My example query:

```graphql
query {
  resourceA(id: ...) {
    resourceB {
      resourceC {
        id
      }
    }
  }
}
```

My assumption and expectation are that each `authorize` of the **respective** DTO is called (i.e. it cascades down).
So, first the `authorize` of `ResourceADto`, then `ResourceBDto`, then `ResourceCDto`.

For the first "level", this works. `ResourceADto` is called. Then the authorizer for `ResourceBDto`.
But `ResourceCDto` is never called, leaving it open.

What I have debugged so far is, that it seems that `DefaultAuthorizer.authorizeRelation(relationName = 'resourceC')` gets called on the `ResourceADto` and then it only searches for the `resourceC`-relation on the "root DTO" of the query (i.e. only on `resourceA`).
It doesn't search for the authorizer on `ResourceBDto` (which, IMHO would be the correct place).

I suspect that in this area lies the problem (not 100% sure since I don't know this codebase all too well):
https://github.com/doug-martin/nestjs-query/blob/master/packages/query-graphql/src/auth/default-crud.authorizer.ts#L67-L68

But I'm missing some kind of "recursive lookup". It only looks for the relation names on "its own DTO-level" and doesn't 1) get the respective relation, then 2), look for the relation on that relation DTO and so on (i.e. it does not cascade down and hence won't find the authorizer for `ResourceC`)

I hope I could describe my assumption/expectation clearly enough.

Maybe I'm wrong here? What do you think?

**Comment by steinroe** [comment]
I think you are right - thanks for digging this up! Have you debugged what DTO and relationDTO and which Authoriser is used here @lukasender? Would be interesting to see how the interceptors are executed.
https://github.com/doug-martin/nestjs-query/blob/02ed9ed56d8430a7a8e608c5aea2d081e57b0c5f/packages/query-graphql/src/resolvers/relations/read-relations.resolver.ts#L34-L56

From a first glance, I would guess the main issue is that interceptors are applied "globally" and (maybe) not chained correctly? If this is the case, the authoriser is applied for the top-level DTO only.
https://github.com/doug-martin/nestjs-query/blob/02ed9ed56d8430a7a8e608c5aea2d081e57b0c5f/packages/query-graphql/src/interceptors/authorizer.interceptor.ts#L17

If I am correct, I see the following two options:
#### Option 1
If we replace the authoriser of the relation with its own, we would loose the ability to overwrite the relation authorisation
https://github.com/doug-martin/nestjs-query/blob/02ed9ed56d8430a7a8e608c5aea2d081e57b0c5f/packages/query-graphql/src/resolvers/relations/read-relations.resolver.ts#L36-L39
where line 39 would become `{ interceptors: [AuthorizerInterceptor(relationDTO)] },`

#### Option 2
Within the authoriser itself, we could make the following method recursive.
https://github.com/doug-martin/nestjs-query/blob/02ed9ed56d8430a7a8e608c5aea2d081e57b0c5f/packages/query-graphql/src/auth/default-crud.authorizer.ts#L71-L84
But we would have to make sure that we don't run into an infinite loop due to circular relations and make sure that we don't do that every time. Also, there is a possibility that we overwrite the authorisations if a relationName is used multiple times. Maybe we can add the relations with `CurrentDTOName.relationName` to the map to make sure we don't mess things up.

What do you think @lukasender and @doug-martin ?

**Comment by steinroe** [comment]
@lukasender do you have a repo with multi-level relations to test this out?

**Comment by lukasender** [comment]
@steinroe As far as I can tell, it's only ever the "root-DTO" and its authorizer (in my example, the one of `ResourceA`).

If it then tries to authorize relations, it then finds the correct Authorizer/DTO for `resourceB` and cascades down correctly, but not for `resourceC` (since that's not a "direct relation of `ResourceA`).

What would be correct here (IMHO), is that the Authorizer of `ResourceBDto` should get called and authorize its relations (i.e. recursively cascade down).
What the best option for this is, I don't know yet. But I guess, this would be the most robust one. You'd also don't need to prevent conflicts of "relationName"s (as one DTO & Authorizer only handles their own relations).

Not sure if infinitely circular relations would be a problem, as you can't write infinitely circular GraphQL queries. At some point, the depth of the GraphQL query ends. But of course, writing recursive code can produce infinite loops if the "stop condition" is not met/written correctly.

> do you have a repo with multi-level relations to test this out?

I don't yet have a repo with multi-level relations that I can show, unfortunately. But I'll try to create an example over the next few days.

**Comment by lukasender** [comment]
@steinroe @doug-martin As discussed, I've prepared an example and showcase https://github.com/lukasender/nestjs-query/pull/1

I've created the example in the `nestjs-query/examples/auth-depth` (I'm lacking a better name right now) folder and created an `e2e` test.

It basically is a copy of `nestjs-query/examples/auth`. However, I've removed all existing entities and DTOs (except for the `User`) and implemented an example with the structure as [described here](https://github.com/doug-martin/nestjs-query/issues/1344#issuecomment-932077111).

Please check out the [PR description](https://github.com/lukasender/nestjs-query/pull/1#issue-749611088).

Note that it is only a showcase and does not involve a fix.

But I guess it can serve as a foundation for further investigations.

**Comment by steinroe** [comment]
Awesome @lukasender, thanks! I will dig into it this week.

**Comment by doug-martin** [comment]
@steinroe I agree option 2 sounds like the best path to me and would lead to a better end-user experience where "it just works" without additional configuration that could be easily missed. Thank you both for figuring this out and contributing back!

**Comment by steinroe** [comment]
I am a bit under water right now - I will try to get to it by the end of next week!

**Comment by kraenzmiralytik** [comment]
FYI: I struggled several days with the framework not applying authorization as expected on a custom resolvers. Eventually, the following line from this comment https://github.com/doug-martin/nestjs-query/issues/1344#issuecomment-936862663 solved it for me.

```ts
GraphQLModule.forRoot({
    //...
    fieldResolverEnhancers: ['interceptors'],
}),
```

Now authorization is applied as I expected it to.

I still need to understand the consequences of enabling field level enhancers to say whether that's an okay workaround for me or not. Nestjs docs give some input https://docs.nestjs.com/graphql/other-features#execute-enhancers-at-the-field-resolver-level
